fusion.md

```diff
@@ -0,0 +1,1396 @@
+<!-- match a share weight that is a direct child of the sector (if there were -->
+<!-- one)? -->
+
+
+
+### Filter objects
+
+A Filter object allows GCAMFusion to select a subset of a data object that is a
+container for other objects. The available Filters are:
+
+* NoFilter: matches all elements
+* NamedFilter: Calls the -&gt;getName() method on the container to compare in it's predicate. `SIMPLE` and `ARRAY` will never match.
+* YearFilter: Calls the -&gt;getYear() method if the data is `CONTAINER` or uses the `Modeltime` to convert period index to year if the data is `ARRAY` and compares that year in it's predicate. `SIMPLE` data will never match.
+* IndexFilter: Computes the index offset of each element in an `ARRAY` or `CONTAINER` Data for comparison in it's predicate. `SIMPLE` Data will never match.
+
+
+
+### Predicates
+
+
+
+A predicate is a way to test whether a year, or name, or index, etc matches a
+value the user was looking for. Currently predicates can only operate on string
+and int. If a predicate that is doing a string comparison is given an int to
+match (i.e. called from a YearFilter) it will always return false, and vice
+versa. The available predicates are:
+<!-- XXX: Is it true that an int predicate will always fail to match a string? -->
+
+|   StringEquals | string | Tests if the proposition exactly matches a string value.|
+|   StringRegexMatches | string | Tests if the proposition matches a regular expression in the egrep notation.|
+|   IntEquals | int | Tests if the proposition exactly matches an int value.|
+|   IntGreaterThan | int | Tests if the proposition is strictly greater than an int value.|
+|   IntGreaterThanEq | int | Tests if the proposition is greater or equal to an int value.|
+|   IntLessThan | int | Tests if the proposition is strictly less than an int value.|
+|   IntLessThanEq | int | Tests if the proposition is less or equal to an int value.|
+
+<!-- XXX: which notation do we want to use, check C++ standard for options -->
+
+
+### Constructing filters using a string notation
+
+The `parseFilterString` utility allows users to construct filters using a
+convenient text notation, instead of constructing them manually.  The rules for
+constructing the string are:
+
+* Separate steps with `/`.  Each step matches a data name. *E.g.*, `region/sector`. 
+* A filter object can optionally be specified inside square brackets. *E.g.*, `/ghg[NamedFilter,StringEquals,CO2]` 
+  * All characters in a filter step up to the opening bracket assumed to the
+    data name. (`ghg` in the example)
+  * All characters in between the '\[' and '\]' (if they exist) are split by
+    ','.  In the example the elements would be `NamedFilter`, `StringEquals`,
+    and `CO2`.
+    * The first element is the [Filter](#filter-objects)  
+    * Unless the filter is a `NoFilter`, there must be a total of three elements.  
+  <!-- XXX: If the filter is a NoFilter, are extra groupings ignored, or are -->
+  <!-- they an error? -->
```

Comment:
Just like the comment says, if you use an explicit `NoFilter`, what are the requirements on the input string?

Reply:
It will result in an error.  I'll update the documentation to clarify.